src/dune_lang/lib_mode.ml

```diff
@@ -0,0 +1,157 @@
+open! Stdune
+open Ocaml
+
+type t =
+  | Ocaml of Mode.t
+  | Melange
+
+module Cm_kind = struct
+  type melange =
+    | Cmi
+    | Cmj
+
+  type t =
+    | Ocaml of Cm_kind.t
+    | Melange of melange
+
+  let choose ocaml melange = function
+    | Ocaml k -> ocaml k
+    | Melange k -> melange k
+
+  let source =
+    choose Cm_kind.source (function
+      | Cmi -> Ml_kind.Intf
+      | Cmj -> Impl)
+
+  let ext =
+    choose Cm_kind.ext (function
+      | Cmi -> ".cmi"
+      | Cmj -> ".cmj")
+
+  let cmi = function
+    | Ocaml _ -> Ocaml Cmi
+    | Melange _ -> Melange Cmi
+
+  let melange_to_dyn =
+    let open Dyn in
+    function
+    | Cmi -> variant "cmi" []
+    | Cmj -> variant "cmj" []
+
+  let to_dyn =
+    let open Dyn in
+    function
+    | Ocaml k -> variant "ocaml" [ Cm_kind.to_dyn k ]
+    | Melange k -> variant "melange" [ melange_to_dyn k ]
+
+  module Dict = struct
+    type 'a melange =
+      { cmi : 'a
+      ; cmj : 'a
+      }
+
+    type 'a t =
+      { ocaml : 'a Cm_kind.Dict.t
+      ; melange : 'a melange
+      }
+
+    let get t = function
+      | Ocaml k -> Cm_kind.Dict.get t.ocaml k
+      | Melange k -> (
+        match k with
+        | Cmi -> t.melange.cmi
+        | Cmj -> t.melange.cmj)
+
+    let make_all x =
+      { ocaml = Cm_kind.Dict.make_all x; melange = { cmi = x; cmj = x } }
+  end
+end
+
+let equal x y =
+  match (x, y) with
+  | Ocaml a, Ocaml b -> Mode.equal a b
+  | Melange, Melange -> true
+  | Ocaml _, Melange | Melange, Ocaml _ -> false
+
+let choose ocaml melange = function
+  | Ocaml m -> ocaml m
+  | Melange -> melange
+
+let to_string = choose Mode.to_string "melange_experimental"
+
+let encode t = Dune_sexp.Encoder.string (to_string t)
+
+let decode =
+  let open Dune_sexp.Decoder in
+  enum
+    [ ("byte", Ocaml Byte)
+    ; ("native", Ocaml Native)
+    ; ("melange_experimental", Melange)
```

Comment:
The name `melange` will do just fine.

To make a feature as experimental, we should do the following:
1. Introduce a melange "syntax"
2. Set the version of this syntax as pre 1.0 (0.1 for example).
3. Guard this new melange mode behind the new syntax.

If you need more pointers on how to do this, let me know. The coq rules is a good reference on how to do this.

Reply:
Thanks for the explanation. I applied the changes in https://github.com/jchavarri/dune/pull/4/commits/f0b3448473d5bc2f36ba558ba31d6920c7a764f8. I understand no changes are needed on the decoding side? (i.e. `Mode_conf.Lib.Set.decode`)